src/Microsoft.AspNetCore.OData/Routing/ODataActionSelector.cs

```diff
@@ -126,31 +136,51 @@ public ActionDescriptor SelectBestCandidate(RouteContext context, IReadOnlyList<
                 // Method(key,relatedKey) vs Method(key).
                 // Method(key,relatedKey,ODataPath) vs Method(key,relatedKey).
                 var matchedCandidates = considerCandidates
-                    .Where(c => !c.FilteredParameters.Any() || TryMatch(c.FilteredParameters, availableKeys, optionalWrapper))
+                    .Where(c => TryMatch(context, c.FilteredParameters, availableKeys, optionalWrapper, c.TotalParameterCount))
                     .OrderByDescending(c => c.FilteredParameters.Count)
-                    .ThenByDescending(c => c.TotalParameterCount);
+                    .ThenByDescending(c => c.TotalParameterCount)
+                    .ToList();
 
-                // Return either the best matched candidate or the first
-                // candidate if none matched.
                 return (matchedCandidates.Any())
                     ? candidates.Where(c => c.Id == matchedCandidates.FirstOrDefault().Id).FirstOrDefault()
-                    : candidates.FirstOrDefault();
+                    : null;
             }
 
             return _innerSelector.SelectBestCandidate(context, candidates);
         }
 
-        private bool TryMatch(IList<ParameterDescriptor> parameters, IList<string> availableKeys, ODataOptionalParameter optionalWrapper)
+        private bool TryMatch(RouteContext context, IList<ParameterDescriptor> parameters, IList<string> availableKeys, ODataOptionalParameter optionalWrapper, int totalParameterCount)
         {
+            if (totalParameterCount == 0 && availableKeys.Count > 0)
+            {
+                return false;
+            }
```

Comment:
similar to asp.net case -- this seems to special case ==0, and be brittle (assuming we've removed any inapplicable keys).